### ISSUE TITLE
OSH fails to start when built on Void Linux xbps-src

### ISSUE THREAD

**Issue by maxice8** [oilshell/oil#93]
xbps-src installs package to destdir/${pkgname}-${version} and creates a package out of the hirearchy found, making an osh package in such manner causes osh to fail

```
$ env OVM_VERBOSE=1 osh
ovm_path = ovm_path_buf (/destdir//osh-0.4.0/usr/bin/oil.ovm)
ovm_path: /destdir//osh-0.4.0/usr/bin/oil.ovm
# installing zipimport hook
import zipimport # builtin
# installed zipimport hook
FATAL: couldn't import from app bundle '/destdir//osh-0.4.0/usr/bin/oil.ovm' (-1)
Stripping the oil.ovm binary may cause this error.
See https://github.com/oilshell/oil/issues/47
```

template 

```
# Template file for 'osh'
pkgname=osh
version=0.4.0
revision=1
wrksrc="oil-${version}"
makedepends="readline-devel"
short_desc="Bash compatible shell written in python"
maintainer="maxice8 <thinkabit.ukim@gmail.com>"
license="Apache-2.0"
homepage="https://www.oilshell.org/"
distfiles="https://www.oilshell.org/download/oil-${version}.tar.xz"
checksum=42e953b9c1c027a947c684de9ccb5fb58f6b6a9c0edbe86e562bc26a88454631
nostrip=yes

do_configure() {
	# configure_args ignore ${DESTDIR}
	./configure --prefix="${DESTDIR}"/usr
}

do_build() {
	make ${makejobs}
}

do_install() {
	vmkdir usr/bin
	./install
}
```

**Comment by andychu** [comment]
Thanks for trying this!

Do you know if xbps-src strips the binary?  That has been the cause of this error in the past, e.g. on Arch.  Arch build tools strip the binary by default, but there is an option to disable this, which fixed it.

**Comment by maxice8** [comment]
yes xbps-src strips binaries and goes great length (hijacks strip binary) to make sure it is the one to do it, it even cries when pre-stripped binaries are present, but i set `nostrip=yes` (arch equivalent is options=(!strip) ) on the template so it won't strip them.

**Comment by andychu** [comment]
Hm, the `--prefix` to configure should be set to the final installation directory, not to an intermediate directory.  From `do_install`, it looks like that is `/usr/bin` ?

I think you want

    configure --prefix /usr

?


Is `$DESTDIR` really set to `/destdir/`?  That seems a bit odd.

**Comment by maxice8** [comment]
> Hm, the --prefix to configure should be set to the final installation directory, not to an intermediate directory. From do_install, it looks like that is /usr/bin ?

if i set to /usr it breaks the package process, xbps-src packages the hirearchy with files found on "${DESTDIR}"


This tree

```
$ tree masterdir/destdir/
masterdir/destdir/
└── osh-0.4.0
    ├── rdeps
    ├── shlib-requires
    └── usr
        └── bin
            ├── oil.ovm
            └── osh -> oil.ovm
```

results in

```
$ xbps-query --repository=hostdir/binpkgs/osh --files osh
/usr/bin/oil.ovm
/usr/bin/osh -> /usr/bin/oil.ovm
```

> Is $DESTDIR really set to /destdir/? That seems a bit odd.

it is basically what most port package managers do, set DESTDIR to point to a fake root and builds on top of it, arch and alpine uses `$pkgdir`, Void Linux uses `$DESTDIR` 



The AUR people also suffer the same [problem](https://aur.archlinux.org/packages/osh/) if they build it with a fakeroot

```
FATAL: couldn't import from app bundle '/tmp/makepkg/osh/pkg/osh/usr/bin/oil.ovm' (-1)
Stripping the oil.ovm binary may cause this error.
See https://github.com/oilshell/oil/issues/47
```

**Comment by andychu** [comment]
I guess I don't understand why the Oil binary is run out of the temp dir?  Is it for testing?

FWIW, the Arch Linux binary now works fine -- @timetoplatypus may have details.  The fix was to not strip the binary.  I don't think the Arch Linux build process ever needs to run the binary out of the temp dir.

In other words, it seems OK to have a non-working binary in the temp dir, and then when it's moved to the final location, it works.  (And I think there are a lot of programs like this, i.e. they won't be able to find stuff in `/usr/share` unless they are in the right location.)

Although I admit the scheme Oil uses is probably bad and needs to be changed.  I was wary about using GNU-specific elf commands, but I think that is probably better than what's there now.

But for now I hope that there is a fix like there was with Arch.

**Comment by maxice8** [comment]
> I guess I don't understand why the Oil binary is run out of the temp dir? Is it for testing?

it is run out of /usr/bin/osh but ovm_path is /destdir/osh-version/usr/bin/osh, i have no clue why the path used to fakeroot leaks into the final binary.

> Although I admit the scheme Oil uses is probably bad, and there's probably a better way to do it. I was wary about using GNU-specific elf commands, but I think that is probably better than what's there now.

it is extremely weird that --prefix affects the final binary.

**Comment by andychu** [comment]
Hm OK can you tell me how to repro?  i.e. clone the void packages repo, then run a xbps-src command?  

Do you have a commit in your fork of the repo that I can test?

**Comment by andychu** [comment]
Oh and it would also be useful to know if the basic install from source without a package manager works for you on Void:

http://www.oilshell.org/release/0.4.0/doc/INSTALL.html

At least then we have a baseline.

**Comment by maxice8** [comment]
> Hm OK can you tell me how to repro? i.e. clone the void packages repo, then run a xbps-src command?

mostly yes, you will also need XBPS binaries in your path, you can download static binaries here https://www.voidlinux.eu/usage/xbps/#download-static-binaries, i honestly would set a chroot

1. download static binaries and put in your PATH
2. clone the void-packages repo then cd into it
3. `mkdir srcpkgs/osh`
4. copy the template i presented on this issue to `srcpkgs/osh`
5. `./xbps-src pkg osh`
6. `xbps-install --repository=hostdir/binpkgs osh`

**Comment by maxice8** [comment]
> Oh and it would also be useful to know if the basic install from source without a package manager works for you on Void:

yes it works. if i set --prefix /usr and install normally then it works outside of the package manager.

**Comment by timetoplatypus** [comment]
Ya for the Arch package, that build issue was resolved by disabling binary stripping. If that's disabled in the Void package, I doubt it's the culprit here.

At least in the Arch package, I do have it configure OSH using the makepkg fakeroot path (` ./configure --prefix="$pkgdir/usr"`) , not the true/final binary install destination.

**Comment by timetoplatypus** [comment]
That OSH builds fine outside of the package manager is what strikes me as the most odd.
Disclaimer: I don't know anything about the Void Linux package manager and I didn't write the configure script, so my commentary may be of limited value.

If all that's changing between builds is the use of the Void Linux package manager, perhaps the environment it compiles software in may not be what we expect. If you look in the configure script, there are things like C version detection, readline detection, etc. If the Void package manager is entering a temporary build environment, maybe there are things the configure script expected to have that aren't present in the temporary environment (or vice versa), and that's causing `./configure` to take an execution path we didn't expect?

**Comment by andychu** [comment]
@maxice8 Thanks for trying it.  I haven't found the time to repro yet, since I have a bunch of other things on my TODO list.

But this is good feedback that the app bundle format is not working well.  The work I'm doing now [1] should resolve this, but it will take awhile (at least a few months, because it's a large change, not just about the app bundles).  In short I don't think I will be relying on `import.c` in Python, so the `$PYTHONPATH` issues should go away.  What's there is pretty hacky.

So I will just leave this issue open, but it might open for awhile :-/  Thanks again.

[1] http://www.oilshell.org/blog/2018/03/04.html
